Mathlib/Data/Nondet/Basic.lean

```diff
@@ -0,0 +1,178 @@
+
+/-- Convert any value in the monad to the singleton nondeterministic value. -/
+def singletonM (x : m α) : Nondet m α :=
+  .mk <| .singletonM do
+    let a ← x
+    return (← saveState, a)
+
+/-- Convert a value to the singleton nondeterministic value. -/
+def singleton (x : α) : Nondet m α := singletonM (pure x)
+
+/-- `Nondet m` is a monad. -/
+instance : Monad (Nondet m) where
+  pure a := singletonM (pure a)
+  bind := bind
+
+/-- `Nondet m` is an alternative monad. -/
+instance : Alternative (Nondet m) where
+  failure := .nil
+  orElse x y := .mk <| x.toMLList.append fun _ => (y ()).toMLList
+
+instance : MonadLift m (Nondet m) where
+  monadLift := singletonM
+
+/--
+Lift a list of monadic values to a nondeterministic value.
+We ensure that each monadic value is evaluated with the same backtrackable state.
+-/
+def ofListM (L : List (m α)) : Nondet m α :=
+  .squash fun _ => do
+    let s ← saveState
+    return .mk <| MLList.ofMLList <| L.map fun x => do
+      restoreState s
+      let a ← x
+      pure (← saveState, a)
+
+/--
+Lift a list of values to a nondeterministic value.
+(The backtrackable state in each will be identical:
+whatever the state was when we first read from the result.)
+-/
+def ofList [Alternative m] (L : List α) : Nondet m α := ofListM (L.map pure)
+
+/-- Apply a function which returns values in the monad to every alternative of a `Nondet m α`. -/
+partial def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
+  L.bind fun a => singletonM (f a)
+
+/-- Apply a function to each alternative in a `Nondet m α` . -/
+def map (f : α → β) (L : Nondet m α) : Nondet m β :=
+  L.mapM fun a => pure (f a)
+
+/-- Convert a monadic optional value to a nondeterministic value. -/
+def ofOptionM (x : m (Option α)) : Nondet m α := .squash fun _ => do
+  match ← x with
+  | none => return .nil
+  | some a => return singleton a
+
+/-- Convert an optional value to a nondeterministic value. -/
+def ofOption (x : Option α) : Nondet m α := ofOptionM (pure x)
+
+/-- Filter and map a nondeterministic value using a monadic function which may return `none`. -/
+partial def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
+  L.bind fun a => ofOptionM (f a)
+
+/-- Filter and map a nondeterministic value. -/
+def filterMap (f : α → Option β) (L : Nondet m α) : Nondet m β :=
+  L.filterMapM fun a => pure <| (f a)
+
+/-- Filter a nondeterministic value using a monadic predicate. -/
+partial def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
+  L.filterMapM fun a => do
+    if (← p a).down then
+      pure (some a)
+    else
+      pure none
+
+/-- Filter a nondeterministic value. -/
+def filter (p : α → Bool) (L : Nondet m α) : Nondet m α :=
+  L.filterM fun a => pure <| .up (p a)
+
+/--
+Find the first alternative in a nondeterministic value, as a monadic value.
+-/
+def head [Alternative m] (L : Nondet m α) : m α := (·.2) <$> MLList.head L.toMLList
```

Comment:
Could you verify that dropping the state here is actually what you want? Given my current understanding I might have thought you would want to restore the state associated to the value:
```suggestion
def head [Alternative m] (L : Nondet m α) : m α := do
  let (s, x) ← L.toMLList.head
  restoreState s
  return x
```